stan/math/rev/mat/functor/map_rect_concurrent.hpp

```diff
@@ -0,0 +1,101 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_MAP_RECT_CONCURRENT_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_MAP_RECT_CONCURRENT_HPP
+
+#include <stan/math/prim/mat/fun/typedefs.hpp>
+
+#include <stan/math/prim/mat/functor/map_rect_reduce.hpp>
+#include <stan/math/prim/mat/functor/map_rect_concurrent.hpp>
+#include <stan/math/prim/scal/err/invalid_argument.hpp>
+#include <stan/math/rev/mat/functor/map_rect_reduce.hpp>
+#include <stan/math/rev/core/chainablestack.hpp>
+
+#include <vector>
+#include <thread>
+#include <future>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+template <int call_id, typename F, typename T_shared_param,
+          typename T_job_param>
+Eigen::Matrix<typename stan::return_type<T_shared_param, T_job_param>::type,
+              Eigen::Dynamic, 1>
+map_rect_concurrent(
+    const Eigen::Matrix<T_shared_param, Eigen::Dynamic, 1>& shared_params,
+    const std::vector<Eigen::Matrix<T_job_param, Eigen::Dynamic, 1>>&
+        job_params,
+    const std::vector<std::vector<double>>& x_r,
+    const std::vector<std::vector<int>>& x_i, std::ostream* msgs) {
+  typedef map_rect_reduce<F, T_shared_param, T_job_param> ReduceF;
+  typedef map_rect_combine<F, T_shared_param, T_job_param> CombineF;
+
+  const int num_jobs = job_params.size();
+  const vector_d shared_params_dbl = value_of(shared_params);
+  std::vector<std::future<std::vector<matrix_d>>> futures;
+
+  auto execute_chunk = [&](int start, int size) -> std::vector<matrix_d> {
+    const int end = start + size;
+    init();
```

Comment:
Is the only change in this file the addition of this init(); call?

Reply:
Yes. Nothing more. Ok, I also need to include the `chainablestack.hpp`, but that's it.